### DISCUSSION TITLE
A question about `cleanupEffect` in reactivity/effect.ts

### DISCUSSION THREAD

**Discussion by SnowingFox** [vuejs/rfcs#442] [General]
when i read effect.ts, i found this function
```ts
function cleanupEffect(effect: ReactiveEffect) {
  const { deps } = effect
  if (deps.length) {
    for (let i = 0; i < deps.length; i++) {
      deps[i].delete(effect)
    }
    deps.length = 0
  }
}
```

I have a question, why should we clean deps? 
```js
for (let i = 0; i < deps.length; i++) {
    deps[i].delete(effect)
}
```
I found the `deps` type declaration is a similarly to `Set[]`, so why not set the deps length to `0` which also can clear deps item like this
```js
deps.length = 0
```

**Comment by jods4** [comment]
You need to clean both ways.

An effect has a `deps: Dep[]`, which is a list of dependencies that can trigger the effect.
Each dependency is a `Set<ReactiveEffect>`, that contains all effects that are triggered when the dependency changes.

Note that dependencies can link to other (multiple) effects that also depend on it.

So when you want to disconnect all dependencies from an effect, you need two steps:

1. Remove the effect from each dependency:
```ts
for (let i = 0; i < deps.length; i++) {
    deps[i].delete(effect)
}
```

2. Remove every dependency from the effect:
```ts
deps.length = 0
```

**Reply by SnowingFox** [reply]
thanks, it's great for me to understand it😄
